src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java

```diff
@@ -543,63 +526,66 @@ public boolean getShowGitToolOptions() {
          * @throws FormException if any data is missing
          */
         @Override
-        public SCM newInstance(@Nullable StaplerRequest req, JSONObject formData)
-                throws FormException {
-            try {
-                BitbucketSCM scm = (BitbucketSCM) super.newInstance(req, formData);
-                scm.setBitbucketClientFactoryProvider(bitbucketClientFactoryProvider);
-                scm.setBitbucketPluginConfiguration(bitbucketPluginConfiguration);
-
-                String serverId = formData.getString("serverId");
-                BitbucketServerConfiguration serverConf = bitbucketPluginConfiguration.getServerById(serverId)
-                        .orElseThrow(() -> new FormException("No server with ID " + serverId, "serverId"));
-
-                String credentialsId = formData.getString("credentialsId");
-                BitbucketCredentials credentials = createWithFallback(CredentialUtils.getCredentials(credentialsId), serverConf);
-                BitbucketClientFactory clientFactory = bitbucketClientFactoryProvider.getClient(serverConf.getBaseUrl(), credentials);
-
-                String projectName = formData.getString("projectName");
-                if (isBlank(projectName)) {
-                    throw new BitbucketSCMException("A project is required", "projectName");
-                }
+        public SCM newInstance(@Nullable StaplerRequest req, JSONObject formData) throws FormException {
+            BitbucketSCM scm = (BitbucketSCM) super.newInstance(req, formData);
+            scm.setBitbucketClientFactoryProvider(bitbucketClientFactoryProvider);
+            scm.setBitbucketPluginConfiguration(bitbucketPluginConfiguration);
+
+            String serverId = formData.getString("serverId");
+            if (isBlank(serverId)) {
+                LOGGER.severe("Error creating the Bitbucket SCM: The serverId cannot be blank");
+                return scm;
+            }
+            Optional<BitbucketServerConfiguration> maybeServerConf = bitbucketPluginConfiguration.getServerById(serverId);
+            if (!maybeServerConf.isPresent()) {
+                LOGGER.severe("Error creating the Bitbucket SCM: No server configuration for the given server id " + serverId);
+                return scm;
+            }
+            BitbucketServerConfiguration serverConf = maybeServerConf.get();
+
+            String credentialsId = formData.getString("credentialsId");
```

Comment:
```suggestion
            String credentialsId = scm.getCredentialsId();
```
I don't think we should interact with the form if we can avoid it, this way we get all the type safety. If we were to rename this field, this method would still work as the refactoring would rename the getter in here as well. We do not have to remember to change a magic string.

Reply:
The credential never gets saved as a field in the constructor. If you look at the implementation of `getCrednetialsId` it gets it off the repository that hasn't been set yet (`getBitbucketSCMRepository().getCredentialsId()`).